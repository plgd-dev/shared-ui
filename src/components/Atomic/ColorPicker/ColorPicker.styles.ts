import { css } from '@emotion/react'
import { ThemeType, get } from '../_theme'

export const colorPicker = (theme: ThemeType) => css`
    display: inline-flex;
    align-items: center;
    padding: 4px;
    background: ${get(theme, `ColorPicker.background`)};
    border-radius: 8px;
    cursor: pointer;
    border: 1px solid ${get(theme, `ColorPicker.borderColor`)};
`

export const color = css`
    width: 40px;
    height: 16px;
    border-radius: 8px;
`

export const hex = css`
    width: 115px;
`

export const rgba = css`
    width: 180px;
`

export const label = (theme: ThemeType) => css`
    font-size: 12px;
    display: block;
    padding: 0 8px;
    color: ${get(theme, `ColorPicker.label.color`)};
`

export const floatingMenu = (theme: ThemeType) => css`
    z-index: 10;
    border: 1px solid ${get(theme, `ColorPicker.floatingMenu.borderColor`)};
    box-shadow: 0 30px 40px rgba(28, 52, 99, 0.1);
    border-radius: 8px;
    overflow: hidden;
`
