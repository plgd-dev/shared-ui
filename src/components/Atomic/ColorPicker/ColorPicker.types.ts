import { ReactNode, SyntheticEvent } from 'react'
import { Placement } from '@floating-ui/react-dom-interactions'
import { RGBColor } from 'react-color'

export type Props = {
    className?: string
    dataTestId?: string
    defaultColor?: RGBColor | string
    id?: string
    menuProps?: {
        placement?: Placement
    }
    onToggle?: (isOpen: boolean, event: SyntheticEvent<any>) => void
    onColorChange?: (color: string) => void
    portalTarget?: ReactNode | Element | null
}

export const defaultProps = {}
