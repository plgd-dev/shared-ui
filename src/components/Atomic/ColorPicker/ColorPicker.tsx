import { FC, useCallback, useEffect, useMemo, useRef, useState } from 'react'
import { createPortal } from 'react-dom'
import { ChromePicker, ColorResult, RGBColor } from 'react-color'
import { offset, shift, useFloating } from '@floating-ui/react'
import { useTheme } from '@emotion/react'
import isFunction from 'lodash/isFunction'

import { Props, defaultProps } from './ColorPicker.types'
import * as styles from './ColorPicker.styles'
import { get } from '../_theme'
import { rgbToHex, hexToRgbA } from '../_utils/commonStyles'
import { colors } from '../_utils/colors'

const ColorPicker: FC<Props> = (props) => {
    const { defaultColor: defaultColorProps, className, dataTestId, id, menuProps, portalTarget, onToggle, onColorChange } = { ...defaultProps, ...props }
    const defaultColor: RGBColor = useMemo(() => {
        if (typeof defaultColorProps === 'string') {
            if (defaultColorProps.length === 7) {
                // hex color
                return hexToRgbA(defaultColorProps, 1, true) as RGBColor
            } else {
                // rgba color
                const parsed = defaultColorProps.split('(')[1].split(')')[0].split(',')
                return {
                    r: parseInt(parsed[0]),
                    g: parseInt(parsed[1]),
                    b: parseInt(parsed[2]),
                    a: parseFloat(parsed[3]),
                }
            }
        } else if (defaultColorProps) {
            return defaultColorProps
        }

        return hexToRgbA(colors.primary, 1, true) as RGBColor
    }, [defaultColorProps])

    const [color, setColor] = useState<RGBColor>(defaultColor)
    const [open, setOpen] = useState(false)

    useEffect(() => {
        setColor(defaultColor)
    }, [defaultColor])

    const { x, y, refs, strategy } = useFloating({
        placement: menuProps?.placement || 'bottom-start',
        strategy: 'fixed',
        middleware: [shift(), offset(4)],
    })

    const ref = useRef(null)
    const theme = useTheme()

    const getValue = useCallback((c: RGBColor) => (c.a !== 1 ? `rgba(${c.r},${c.g},${c.b},${c.a})` : rgbToHex(c.r, c.g, c.b)), [])

    const handleChange = useCallback(
        (color: ColorResult) => {
            setColor(color.rgb)
            isFunction(onColorChange) && onColorChange(getValue(color.rgb))
        },
        [getValue, onColorChange]
    )

    useEffect(() => {
        function handleClickOutside(event: any) {
            // @ts-ignore
            if (ref?.current && !ref?.current?.contains(event.target)) {
                setOpen(false)
                isFunction(onToggle) && onToggle(false)
            }
        }
        if (open) {
            document.addEventListener('mousedown', handleClickOutside)
            return () => {
                document.removeEventListener('mousedown', handleClickOutside)
            }
        }
    }, [ref, open, onToggle])

    const FloatingPanel = (
        <div
            css={styles.floatingMenu}
            ref={refs.setFloating}
            style={{
                position: strategy,
                top: y ?? 0,
                left: x ?? 0,
                width: 'max-content',
            }}
        >
            <ChromePicker
                className='color-picker'
                color={color}
                onChange={handleChange}
                styles={{
                    default: {
                        body: {
                            background: get(theme, `ColorPicker.background`),
                        },
                    },
                }}
            />
        </div>
    )

    return (
        <div className={className} id={id} ref={ref}>
            <div
                css={[styles.colorPicker, getValue(color).startsWith('#') ? styles.hex : styles.rgba]}
                data-test-id={dataTestId}
                onClick={() => setOpen((prev) => !prev)}
            >
                <div
                    css={styles.color}
                    style={{
                        background: getValue(color),
                    }}
                />
                {color ? <span css={styles.label}>{getValue(color)}</span> : null}
            </div>
            {open && portalTarget && createPortal(FloatingPanel, portalTarget as Element)}
            {open && !portalTarget && FloatingPanel}
        </div>
    )
}

ColorPicker.displayName = 'ColorPicker'

export default ColorPicker
