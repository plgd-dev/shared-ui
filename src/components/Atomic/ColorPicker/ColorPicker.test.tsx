import { render } from '../../../../test/jest-wrapper'
import ColorPicker from './ColorPicker'

describe('<ColorPicker>', () => {
    it('renders without crashing', () => {
        const { container, asFragment } = render(<ColorPicker />)
        expect(container).toBeInTheDocument()

        expect(asFragment()).toMatchSnapshot()
    })
})
