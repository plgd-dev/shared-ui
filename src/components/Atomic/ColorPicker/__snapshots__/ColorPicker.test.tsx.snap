// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ColorPicker> renders without crashing 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 4px;
  background: #FFFFFF;
  border-radius: 8px;
  cursor: pointer;
  border: 1px solid #E6E9ED;
  width: 115px;
}

.emotion-1 {
  width: 40px;
  height: 16px;
  border-radius: 8px;
}

.emotion-2 {
  font-size: 12px;
  display: block;
  padding: 0 8px;
  color: #81868C;
}

<div>
    <div
      class="emotion-0"
    >
      <div
        class="emotion-1"
        style="background: rgb(34, 97, 174);"
      />
      <span
        class="emotion-2"
      >
        #2261ae
      </span>
    </div>
  </div>
</DocumentFragment>
`;
